File: Project Euler #14.py
import time


def longest_collatz_sequence(upper_bound):
    chain, record = 0, 0
    numbers = {}
    for x in range(3, upper_bound, 2):
        y = x
        if chain > record:
            record2, record = y-2, chain
        chain = 0
        while x > 1:
            if x % 2 == 0:
                x /= 2
                chain += 1
            else:
                if x in numbers:
                    chain += numbers[x]
                    break
                x = (3*x+1)/2
                chain += 2
        numbers[y] = chain
    print(record2)


if __name__ == "__main__":
    start = time.time()
    longest_collatz_sequence(1000000)
    end = time.time()
    print(end - start)

File: Project Euler #14.cpp
#include <iostream>
#include<ctime> 
using namespace std;
int chain = 0;
int record = 0;
int record_x_value = 0;
int start_s=clock() ;

void longest_collatz_sequence(int upper_bound){
    for (int x = 3; x < upper_bound; x += 2) {
        long long y = x;
        if (chain > record) {
            record = chain;
            record_x_value = x;}
        chain = 0;
        while (y > 1) {
            if (y % 2 == 0) {
                y /= 2;
                chain++;}
            else {
                y = (3 * y + 1)/2;
                chain += 2;}}}
}

int main()
{
   longest_collatz_sequence(1000000);
    cout << record_x_value << endl;
    int stop_s=clock();
    cout<<(stop_s-start_s)/double(CLOCKS_PER_SEC);
}
